File: src/#1305/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getAllElements(root1 *TreeNode, root2 *TreeNode) []int {
	arr1 := []int{}
	arr2 := []int{}

	if root1 == nil && root2 != nil {
		traverce(root2, &arr2)
		return arr2
	}
	if root1 != nil && root2 == nil {
		traverce(root1, &arr1)
		return arr1
	}

	traverce(root1, &arr1)
	traverce(root2, &arr2)

	i, j := 0, 0
	res := []int{}
	for i < len(arr1) && j < len(arr2) {
		if arr1[i] < arr2[j] {
			res = append(res, arr1[i])
			i++
		} else {
			res = append(res, arr2[j])
			j++
		}
	}

	if i == len(arr1) {
		res = append(res, arr2[j:]...)
	} else {
		res = append(res, arr1[i:]...)
	}
	
	return res
}
func traverce(root *TreeNode, arr *[]int) {
	if root == nil {
		return
	}
	traverce(root.Left, arr)
	*arr = append(*arr, root.Val)
	traverce(root.Right, arr)
}

File: src/#1305/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var getAllElements = function (root1, root2) {
  const values1 = traverse(root1, [])
  const values2 = traverse(root2, [])

  let ans = []
  let i = 0,
    j = 0
  while (i < values1.length && j < values2.length) {
    if (values1[i] < values2[j]) {
      ans.push(values1[i])
      i++
    } else {
      ans.push(values2[j])
      j++
    }
  }
  if (i < values1.length) {
    ans = ans.concat(values1.slice(i))
  } else {
    ans = ans.concat(values2.slice(j))
  }
  return ans

  // in-order traversal
  function traverse(root, values) {
    if (!root) return values
    traverse(root.left, values)
    values.push(root.val)
    traverse(root.right, values)
    return values
  }
}
